### PR TITLE
Reflectors

### PR DESCRIPTION
**Pull request by StanberyTrask**
## About the PR 
Adds Reflectors, still needs work with reflection logic but other than that they're fully implemented with a full construction tree and angle rotation by clicking on the reflector

**Screenshots**

Screenshot from 2021-11-30 21-44-05
**Changelog**

:cl: Stanbery Trask
- add: Adds Single, Double, and Box Reflectors
- add: Added a construction tree to reflectors

### CODE REVIEW

File: Content.Server/Reflector/ReflectorComponent.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        [DataField("Angle")]
+        public float Angle { get; set; } = 0;
+
+        /// <summary>
+        ///     If the Reflector is one sided(0), two sided(1), or a box Reflector(2)
+        /// </summary>
+        [ViewVariables]
+        [DataField("Type")]
```
Comment (mirrorcult) [inline Content.Server/Reflector/ReflectorComponent.cs:30]: just make it an enum?

File: Content.Server/Reflector/ReflectorComponent.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace Content.Server.Reflector
+{
+    [RegisterComponent, Friend(typeof(ReflectorSystem))]
+    public class ReflectorComponent : Component
+    {
+        public override string Name => "Reflector";
+
+        [DataField("whitelist")]
+        public EntityWhitelist Whitelist = new();
```
Comment (mirrorcult) [inline Content.Server/Reflector/ReflectorComponent.cs:17]: make this a required field

File: Content.Server/Reflector/ReflectorSystem.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            if (EntityManager.TryGetComponent<SpriteComponent>(args.TargetUid, out SpriteComponent? sprite))
+            {
+                sprite.LayerSetRotation(1, ToRadians(reflector.Angle));
+            }
+
+            //so someone cant sit there for an hour clicking it to crash the server
+            if (reflector.Angle >= 360)
+            {
+                reflector.Angle -= 360;
+            }
```
Comment (mirrorcult) [inline Content.Server/Reflector/ReflectorSystem.cs:36]: why not just use rotatable?

File: Resources/Locale/en-US/emitter-mirror/reflector-system.ftl [hunk 4]
```diff
@@ -0,0 +1 @@
+reflector-component-on-examine-message = A
```
Comment (mirrorcult) [inline Resources/Locale/en-US/emitter-mirror/reflector-system.ftl:1]: a

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Content.Server/Reflector/ReflectorComponent.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+        /// <summary>
+        ///     The Angle of the reflector
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        [DataField("Angle")]
+        public float Angle { get; set; } = 0;
```
Comment (metalgearsloth) [inline Content.Server/Reflector/ReflectorComponent.cs:24]: This should be using the entity's rotation instead.

File: Resources/Locale/en-US/emitter-mirror/reflector-system.ftl (hunk 4 above)
Comment (Efruit) [reply Resources/Locale/en-US/emitter-mirror/reflector-system.ftl:1]: 🅰️

**Pull request closed without merging**